### ISSUE TITLE
[Enhancement] Pipeline placement

### ISSUE THREAD

**Issue by hogura99** [USC-NSL/DisagMoE#26]
Parameters: p=step_attn, q=step_exp

Let Attn_i and Expert_j be the attention for layer i and **all experts** for layer j respectively.

First we get virtual mapping:

```
V_0:        [Attn_0, Attn_p, Attn_{2p}, ...]
V_1:        [Attn_1, Attn_{p+1}, Attn_{2p+1}, ...]
...
V_{p-1}:    [Attn_{p-1}, Attn_{2p-1}, Attn_{3p-1}, ...]

V_p:        [Expert_0, Expert_q, Expert_{2q}, ...]
V_{p+1}:    [Expert_1, Expert_{q+1}, Expert_{2q+1}, ...]
...
V_{p+q-1}:  [Expert_{q-1}, Expert_{2q-1}, Expert_{3q-1}, ...]
```

For each virtual group V_k, we use TP or EP inside the group.

Then we have:
    [V_0, V_{p-1}] -> [G_0, G_{p * TP_SIZE - 1}]
    [V_p, V_{p+q-1}] -> [G_{p * TP_SIZE}, G_{p * TP_SIZE + q * EP_SIZE - 1}]

where `G_i` indicates the actual GPU with id `i`.

Ideally, we still need a further physical mapping to minimize cross-node communication.

**Comment by hogura99** [comment]
#24

**Comment by hogura99** [comment]
Now support `get_model_placement(model_config, cluster_config, "pipeline", step_attn=4, step_expert=1, zigzag_attn=True)`, where `zigzag_attn=True or False`, `step_attn>=1`.
I have only tested `step_expert=1`, don't know if any errors might happen when `step_expert>1`.

**Comment by hogura99** [comment]
Closed by #25
